probe-rs/src/architecture/esp_common.rs

```diff
@@ -0,0 +1,351 @@
+    const USER_MISO: u32 = 1 << 28;
+    const USER_COMMAND: u32 = 1 << 31;
+
+    // user2 register
+    const USER_COMMAND_BITLEN: u32 = 28;
+
+    // miso dlen register
+    const MISO_BITLEN: u32 = 0;
+
+    // cmd register
+    const USER_CMD: u32 = 1 << 18;
+
+    interface.write_word_32(regs.ctrl(), old_ctrl_reg | CTRL_WP)?;
+    interface.write_word_32(regs.user(), old_user_reg | USER_COMMAND | USER_MISO)?;
+    interface.write_word_32(regs.user1(), 0)?;
+    interface.write_word_32(regs.user2(), (7 << USER_COMMAND_BITLEN) | command as u32)?;
+    interface.write_word_32(regs.addr(), 0)?;
+    interface.write_word_32(
+        regs.miso_dlen(),
+        (miso_bits.saturating_sub(1)) << MISO_BITLEN,
+    )?;
+    interface.write_word_32(regs.data_buf_0(), 0)?;
+
+    // Execute read
+    interface.write_word_32(regs.cmd(), USER_CMD)?;
+    while interface.read_word_32(regs.cmd())? & USER_CMD != 0 {}
+
+    // Read result
+    let value = interface.read_word_32(regs.data_buf_0())?;
+
+    // Restore registers
+    interface.write_word_32(regs.ctrl(), old_ctrl_reg)?;
+    interface.write_word_32(regs.user(), old_user_reg)?;
+    interface.write_word_32(regs.user1(), old_user1_reg)?;
+
+    Ok(value)
+}
+
+fn detect_flash_size(
+    interface: &mut impl MemoryInterface,
+    spiflash_addr: u32,
+) -> Result<Option<usize>, crate::Error> {
+    const RDID: u8 = 0x9F;
+
+    let value = execute_flash_command_generic(
+        interface,
+        &SpiRegisters {
+            base: spiflash_addr,
+            cmd: 0x00,
+            addr: 0x04,
+            ctrl: 0x08,
+            user: 0x18,
+            user1: 0x1C,
+            user2: 0x20,
+            miso_dlen: 0x28,
+            data_buf_0: 0x58,
+        },
+        RDID,
+        24,
+    )?;
+
+    Ok(decode_flash_size(value))
+}
+
+fn detect_flash_size_esp32(
+    interface: &mut impl MemoryInterface,
+    spiflash_addr: u32,
+) -> Result<Option<usize>, crate::Error> {
+    const RDID: u8 = 0x9F;
+
+    let value = execute_flash_command_generic(
+        interface,
+        &SpiRegisters {
+            base: spiflash_addr,
+            cmd: 0x00,
+            addr: 0x04,
+            ctrl: 0x08,
+            user: 0x1C,
+            user1: 0x20,
+            user2: 0x24,
+            miso_dlen: 0x2C,
+            data_buf_0: 0x80,
+        },
+        RDID,
+        24,
+    )?;
+
+    Ok(decode_flash_size(value))
+}
+
+fn decode_flash_size(value: u32) -> Option<usize> {
+    let [manufacturer, memory_type, capacity, _] = value.to_le_bytes();
+
+    tracing::debug!(
+        "Detected manufacturer = {:x} memory_type = {:x} capacity = {:x}",
+        manufacturer,
+        memory_type,
+        capacity
+    );
+
+    const KB: usize = 1024;
+    const MB: usize = 1024 * KB;
+
+    // TODO: replace with `espflash::flasher::FlashSize::from_detected` when
+    // https://github.com/esp-rs/espflash/pull/530 gets released.
+    let capacity = match (manufacturer, memory_type, capacity) {
```

Comment:
I don't like where this is going. We need a better system for this. One of the principles of probe-rs is to not make a mess with 1000 special cases like OpenOCD. Having custom mechanisms for every arch is not really nice :/

Reply:
I don't really have a better idea than using the mechanism the external flash ICs provide us with.

Reply:
Also note that we already do this for the RISC-V ESP32s. :)

Reply:
Yeah, I was more speaking of having an actual trait which we can impl for different systems and having a way to select the correct detection algo.

Reply:
This detection is done through the current DebugSequence mechanism, so technically we're there. :) From my perspective it looks weird because the chip-specific code is places in the same module as the architecture-specific implementations are. I'm thinking about moving the debug sequences & all to a new, chip-specific module so that this common code wouldn't be so out of place, but I'm not sure if that sounds good enough.

Reply:
Lets get this merged but iterate on it later.

We really need to take great care to not let it deteriorate ^^ I am constantly torn between more support and more cleanliness :D

Reply:
There's no rush and I'd prefer a solution that doesn't step actively on your toe.

Reply:
@Yatekii could you please take another look how it feels after moving debug sequences into a more target-specific location?